JavaScripts/credit_card_checker.js

```diff
@@ -0,0 +1,191 @@
+// use the Luhn Algorithm to check if a credit card number is valid
+// https://en.wikipedia.org/wiki/Luhn_algorithm#Description
+
+// take every 2nd number from the array (from right to left) and double it
+// last number remains untouched
+// if the value is greater than 9, subtract 9
+// take the sum of all these numbers and modulo by 10
+// if there is a remainder the result is false (ie. card number is not valid)
+
+// abstraction of summing up card numbers
+const summarizeCreditCard = array => {
+  // copy the array, remove last digit and add it to the sum
+  // reverse copied array so the 2nd last digit is now the first
+  let newArray = array.slice(0, -1).toReversed()
+  let sum = array[array.length - 1]
+  // implement math logic for all even indexes starting from 0 
+  newArray.forEach((value, index, array) => {
+    if (index % 2 === 0) {
+      let doubled = value * 2;
+      if (doubled >= 9) {
+        sum += doubled - 9;
+      } else {
+        sum += doubled;
+      }
+    }
+    // odd indexes are simply added as is
+    else {
+      sum += value;
+    }
+  })
+  return sum;
+}
+
+// 1st approach.. got too complicated.. DISCARDED
+const validateCred1 = (array) => {
+  // last value gets added to the sum as it is
+  let sum = array[array.length-1];
+  // iterating backwards starting from the 2nd last position
+  for (let i = array.length-2; i >= 0; i--) {
+console.log(`index ${i}: ${array[i]}`);
+    // only for even indexes the math is applied
+    if (i % 2 === 0) {
+      let doubled = array[i] * 2;
+      if (doubled >= 9) {
+        sum += doubled - 9;
+      } else {
+        sum += doubled;
+      }
+    } 
+    // odd indexes are simply added as is
+    else {
+      sum += array[i];
+    } 
+console.log(`new sum ${sum}`)
+  }
+  if (sum % 10 === 0) {
+    return true;
+  } else {
+    return false;
+  }
+};
+
+// 2nd approach.. (after hint from Codecademy)
+const validateCred = array => {
+  if (summarizeCreditCard(array) % 10 === 0) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// The role of findInvalidCards() is to check through the nested array for which numbers are invalid, and return another nested array of invalid cards.
+const findInvalidCards = (nestedArray) => {
+  return nestedArray.filter((array)=> validateCred(array) === false)
+} // TODO: is there a way to return the variable names only? 
+
+// based on first number will return name of card company of affected invalid credit card numbers
+const idInvalidCardCompanies = (nestedArray) => {
```

Comment:
I'd rename things to be more semantically relevant. 

```
// Ask yourself whether or not it's easy to read and follow the logic of these smaller functions with 
// well chosen function and variable names that say exactly what is happening and what the data is
// relative to the use-case (compared to what you originally had). 
const getCompaniesForInvalidAccounts = (accounts) => {
    const invalidAccounts = accounts.filter(isAccountInvalid);
    const invalidCompanies = invalidAccounts.map(getCompanyForAccount); // Could contain duplicates
    return dedupArray(invalidCompanies);
}

// Defining this small inverse function of isAccountValid means that we can use 
// the very concise and readable code `accounts.filter(isAccountInvalid)` instead of 
// `accounts.filter(account => !isAccountValid(account))` (which is still readable, but the former is just more elegant and cleaner).
const isAccountInvalid = (account) => ! isAccountValid(account); // This is the function I had defined in the comment above.

// Remove duplicates in an array by transforming to a Set and back.
const dedupArray = (array) => [... new Set(array)]; // I named this `array` because this function is completely reusable and generic. We could pull this into an array utilities file in the codebase somewhere if we wanted to. 

// A small helper function that wraps the complexity of a high-level operation behind a simply named function.
const getCompanyForAccount = (account) => {
   const companyInfo = companiesData.find(companyData => account[0] === companyData.accountPrefix);
   return companyInfo?.name ?? 'Unknown';
}

// Keep related data co-located and generalized in an abstract collection (e.g. an array of account information)
// so it's easy to add a new one, since you'll likely need to do so in the future. 
// Your algorithm that uses it should just depend on the generalized abstraction, so when the data changes
// or new companies are added, your algorithm doesn't need to be touched at all. 
// tl;dr: Decouple the data from the algorithm. 
const companiesData = [
     {
         name: 'Amex',
         accountPrefix: 3
      }, 
      {
         name: 'Visa',
         accountPrefix: 4
      },
      {
         name: 'MasterCard',
         accountPrefix: 5
      },
      {
         name: 'Discover',
         accountPrefix: 6
      }
];

Reply:
I tried to refactor my code to a bit more 'senior' approach (trying to use .map and .some with the company data as list of objects) but I just can't wrap my head around it 😩